lib/server.js

```diff
@@ -0,0 +1,198 @@
+var Docker = require('dockerode-promise');
+var debug = require('debug')('docker-exec-websocket-server:lib:server');
+var through = require('through');
+var fs = require('fs');
+var ws = require('ws');
+var slugid = require('slugid');
+var _ = require('lodash');
+var path = require('path');
+var assert = require('assert');
+var url = require('url');
+var msgcode = require('./messagecodes.js');
+
+var dockerSocket = '/var/run/docker.sock';
+var stats = fs.statSync(dockerSocket);
+if (!stats.isSocket()) {
+  throw new Error('Are you sure the docker is running?');
+}
+var docker = new Docker({socketPath: dockerSocket});
+
+// Makes a new buffer with code number (ideally 0x00-0xFF)
+// appended to the start of the given buffer
+function muxBuffer (code, buffer) {
+  return Buffer.concat([new Buffer([code]), buffer]);
+}
+
+/** Makes an instance of docker exec running command, then links the provided
+    websocket to the stream provided by docker exec
+
+    @param tty: boolean representing whether or not terminal emulation is active
+  */
+function execWithSocket (container, socket, command, tty, dbgWrite) {
+  var options = {
+    AttachStdout: true,
+    AttachStderr: true,
+    AttachStdin: true,
+    Tty: tty,
+    Detach: false,
+    Cmd: command,
+  };
+  var attachoptions = {
+    stdin: true,
+    stdout: true,
+    stderr: true,
+    stream: true,
+  };
+  var exec;
+  container.exec(options).then((_exec) => {
+    exec = _exec;
+    return exec.start(attachoptions);
+  }).then((execStream) => {
+    //handling output
+    var strout = through(function (data) {
+      dbgWrite('out', data);
+      var buf = muxBuffer(msgcode.stdout, new Buffer(data));
+      this.queue(buf);
+    });
+
+    var strerr = through(function (data) {
+      dbgWrite('err', data);
+      var buf = muxBuffer(msgcode.stderr, new Buffer(data));
+      this.queue(buf);
+    });
+
+    exec.modem.demuxStream(execStream, strout, strerr);
+    var strbuf = through();
+    strbuf.pipe(through((data) => {
+      socket.send(data, {binary: true});
```

Comment:
`socket.send` takes an optional callback... but not waiting for the callback we just buffer up everything in the network layer...

To avoid this we should ensure we have at lost X bytes (min Y messages) given send callback being called...